fast_llm/data/preparator/gpt_memmap/prepare.py

```diff
@@ -86,25 +98,28 @@ def run(self):
         # Prepare output directory
         self._config.output_path.mkdir(parents=True, exist_ok=True)
 
-        # Download dataset if necessary on rank 0
-        download_path = self._config.output_path / "downloaded_dataset"
-        download_path_ok = download_path / "ok"
-        if self._config.distributed.rank == 0 and not download_path_ok.exists():
-            datasets.load_dataset(
-                path=self._config.dataset.path,
-                name=self._config.dataset.config_name,
-                split=self._config.dataset.split,
-                num_proc=self._config.loading_workers,
-                trust_remote_code=self._config.dataset.trust_remote_code,
-            ).save_to_disk(download_path, num_proc=self._config.saving_workers)
-            download_path_ok.touch()
-
-        # Synchronize processes to wait for the download to finish
-        if self._config.distributed.world_size > 1:
-            torch.distributed.barrier()
+        if pathlib.Path(self._config.dataset.path).is_dir():
+            # Dataset is already downloaded, load from disk
+            dataset = self._load_dataset()
+        else:
+            # Dataset is not downloaded, download on rank 0
+            if self._config.distributed.rank == 0:
+                dataset = self._load_dataset()
+
+            # Synchronize processes to wait for the download to finish on rank 0
+            if self._config.distributed.world_size > 1:
```

Comment:
Is this `if` necessary?
I would expect the barrier to do nothing and pass if `world_size==1`, but I might be wrong.